### PR TITLE
Big fucking refac

### PR DESCRIPTION
**Pull request by Feresey**

### CODE REVIEW

File: init_resolution.go [hunk 1]
```diff
@@ -125,8 +144,10 @@ func getEnabledLeafs(root string,
 			return false
 		}
 	}
+	// ты ради этого result по ссылке передаешь?
```
Comment (Darkclainer) [inline init_resolution.go]: Please, refrain from unnecessary comments.

File: init_resolution.go [hunk 2]
```diff
@@ -144,18 +166,26 @@ func GetDisabledLeafsFromRoot(root string, states map[string]State, requirements
 // F B
 //  /|\
 // C D E
-func getDisabledLeafsFromRoot(root string, states map[string]State, requirements map[string][]string, results *[]string) bool {
+func getDisabledLeafsFromRoot(
+	root string,
+	states map[string]State,
+	requirements map[string][]string,
+	results *[]string) bool {
```
Comment (Darkclainer) [inline init_resolution.go]: Actually, I like this style, but prefer one another thing:
```Go
func getDisabledLeafsFromRoot(
	root string,
	states map[string]State,
	requirements map[string][]string,
	results *[]string,
) bool {
```
This helps extend function arguments easily and makes it easier to see where function argements end and where return values starts.

File: service.go [hunk 3]
```diff
@@ -54,38 +54,45 @@ type Service struct {
...
-		Args:          Args,
+		Name:          name,
+		Command:       command,
+		Args:          args,
 		runningRegexp: runningTemplate,
 	}
 }
 
 func (s *Service) Start(ctx context.Context) chan ServiceMessage {
+	// Ну и нахйя ты передаешь контекст?
```
Comment (Darkclainer) [inline service.go]: Please, refrain from obscene speech in comments.

File: service_manager_test.go [hunk 4]
```diff
@@ -10,49 +10,59 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+const (
+	waitTime = 5 * time.Second
+	aStarted = 1
+)
+
```
Comment (Darkclainer) [inline service_manager_test.go]: No point to hide "magic" numbers in tests. Each test case in slightly unique
But you can hide regexp template "ready" here, it's reinitialized almost in every test case

File: service_manager_test.go [hunk 5]
```diff
@@ -153,41 +172,46 @@ func TestServiceManagerStartWithDependency(t *testing.T) {
...
+	recorded := make([]ServiceMessage, 0, len(messages))
+
 	for message := range messages {
 		switch message.Name {
 		case "A":
 			if message.Type == MessageState && message.State == StateRunning {
-				atomic.AddInt64(&aStarts, 1)
-
+				atomic.AddInt64(&aStarts, aStarted)
```
Comment (Darkclainer) [inline service_manager_test.go]: This is more obscure that it originally was. 
It was a common pattern "increment"

File: service_manager_test.go [hunk 6]
```diff
@@ -153,41 +172,46 @@ func TestServiceManagerStartWithDependency(t *testing.T) {
...
 			}
 		case "B":
 			recorded = append(recorded, message)
-			if message.Type == MessageState && message.State == StateStarted {
-				if atomic.LoadInt64(&aStarts) != 1 {
+
+			if message.Type == MessageState &&
+				message.State == StateStarted {
+				if atomic.LoadInt64(&aStarts) != aStarted {
```
Comment (Darkclainer) [inline service_manager_test.go]: This is more obscure that it originally was.

File: service_manager_test.go [hunk 7]
```diff
@@ -208,20 +232,24 @@ func TestServiceManagerStartWithFullfilledDependency(t *testing.T) {
...
 	m.Start("A")
+
+	recorded := make([]ServiceMessage, 0, len(messages))
+
 	for message := range messages {
 		switch message.Name {
 		case "A":
 			if message.Type == MessageState && message.State == StateRunning {
-				atomic.AddInt64(&aStarts, 1)
+				atomic.AddInt64(&aStarts, aStarted)
```
Comment (Darkclainer) [inline service_manager_test.go]: This is more obscure that it originally was.

File: service_manager_test.go [hunk 8]
```diff
@@ -208,20 +232,24 @@ func TestServiceManagerStartWithFullfilledDependency(t *testing.T) {
...
 				go m.Start("B")
 			}
 		case "B":
 			recorded = append(recorded, message)
-			if message.Type == MessageState && message.State == StateRunning {
-				if atomic.LoadInt64(&aStarts) != 1 {
+
+			if message.Type == MessageState &&
+				message.State == StateRunning {
+				if atomic.LoadInt64(&aStarts) != aStarted {
```
Comment (Darkclainer) [inline service_manager_test.go]: This is more obscure that it originally was.

File: service_manager_test.go [hunk 9]
```diff
@@ -268,7 +295,7 @@ func TestServiceManagerStop(t *testing.T) {
 
 	m.Start("TEST")
 
-	ticker := time.NewTicker(time.Second * 5)
+	ticker := time.NewTicker(waitTime)
```
Comment (Darkclainer) [inline service_manager_test.go]: This is more obscure that it originally was.

File: service_manager_test.go [hunk 10]
```diff
@@ -329,61 +359,87 @@ loop:
 			switch message.Name {
 			case "A":
 				if message.Type == MessageState && message.State == StateRunning {
-					atomic.AddInt64(&aStarts, 1)
+					atomic.AddInt64(&aStarts, aStarts)
```
Comment (Darkclainer) [inline service_manager_test.go]: This is more obscure that it originally was.

File: service_manager_test.go [hunk 11]
```diff
@@ -329,61 +359,87 @@ loop:
...
 				}
+
 				if message.Type == MessageState && message.State == StateFailed {
 					t.Error("aService wasn't stopped gracefully")
 				}
+
 				if message.Type == MessageState && message.State == StateFinished {
-					atomic.AddInt64(&aStops, 1)
+					atomic.AddInt64(&aStops, aStarts)
```
Comment (Darkclainer) [inline service_manager_test.go]: This is more obscure that it originally was.

File: service_manager_test.go [hunk 12]
```diff
@@ -329,61 +359,87 @@ loop:
...
+
 				if message.Type == MessageState && message.State == StateFailed {
 					t.Error("bService wasn't stopped gracefully")
 				}
+
 				if message.Type == MessageState && message.State == StateFinished {
-					assert.Equal(t, int64(1), atomic.LoadInt64(&aStarts), "a wasn't started or was started more than once")
-					assert.Equal(t, int64(1), atomic.LoadInt64(&aStops), "a wasn't stopped or was stopped more than once")
+					assert.Equal(t, aStarts, atomic.LoadInt64(&aStarts), "a wasn't started or was started more than once")
+					assert.Equal(t, aStarts, atomic.LoadInt64(&aStops), "a wasn't stopped or was stopped more than once")
```
Comment (Darkclainer) [inline service_manager_test.go]: This is more obscure that it originally was.

File: service_manager_test.go [hunk 13]
```diff
@@ -329,61 +359,87 @@ loop:
...
+		startTemplate       = regexp.MustCompile("ready")
+		m                   = NewServiceManager()
+		started             = &sync.WaitGroup{}
+		finished            = make(chan struct{})
+		finishes      int64 = 0
+	)
+
+	for _, s := range services {
+		m.Register(s.name, "service", []string{"lines", "ready", "sleep", "5000"}, startTemplate, s.req)
+	}
```
Comment (Darkclainer) [inline service_manager_test.go:417]: I don't think this all refactor for initialisation in "for" was worth it. 
First it is more lenghtly, second now I have much less freedom in how to define object. 
Maybe in production code this is more preferable solution, but in test  it's not.

File: service_manager_test.go [hunk 14]
```diff
@@ -329,61 +359,87 @@ loop:
...
 
-	started := &sync.WaitGroup{}
-	finished := make(chan struct{})
-	var finishes int64 = 0
+	started.Add(len(services))
```
Comment (Darkclainer) [inline service_manager_test.go:419]: I agree that this is less error-prone, but you can do it by just investigating unexported fields of ServiceManager

**Review by Darkclainer — CHANGES_REQUESTED** [review]

File: service_manager_test.go (hunk 13 above)
Comment (Feresey) [reply service_manager_test.go:417]: тут фишка не в for, а в том что инициализация идёт по слайсу тестов. Тогда будет легко добавить тест, просто дописав его в слайс

File: service_manager_test.go (hunk 14 above)
Comment (Feresey) [reply service_manager_test.go:419]: Тогда почему ты сам не добавил проверку по полям структуры?

**Pull request closed without merging**